### PR TITLE
Numerical stability improvements

### PR DESCRIPTION
**Pull request by burgholzer**
⚡ consistent floating point comparisons across all functions
🔥 remove redundant zero check in complex number division
⚡ numerically more stable vector normalization
🐛 ComplexTable::find did not actually correctly iterate over the bucket  
🐛 ComplexTable::insert did not actually correctly insert elements in a sorted fashion
⚡ border cases during complex table lookup can be handled by just looking at two entries
⚡ check for better match during complex table lookup
🎨 improve tolerance checks to also work with 0 tolerance
🐛 also reset complex number table and cache upon reset

### CODE REVIEW

File: include/dd/Definitions.hpp [hunk 1]
```diff
@@ -82,5 +82,19 @@ namespace dd {
     //        return b;
     //    }
 
+    // calculates the Units in Last Place (ULP) distance of two floating point numbers
+    [[maybe_unused]] static std::size_t ulpDistance(dd::fp a, dd::fp b) {
+        if (a == b)
+            return 0;
+
+        std::size_t ulps = 1;
+        dd::fp      iter = std::nextafter(a, b);
```
Comment (hillmich) [inline include/dd/Definitions.hpp:91]: `iter` is not an iterator. Maybe use `nextFp` as variable name?

File: include/dd/ComplexTable.hpp [hunk 2]
```diff
@@ -359,6 +391,7 @@ namespace dd {
         };
 
         void print() {
+            std::cout << std::setprecision(std::numeric_limits<dd::fp>::max_digits10);
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply include/dd/ComplexTable.hpp:394]: Good catch. Thanks 👍🏼

File: include/dd/Definitions.hpp (hunk 1 above)
Comment (burgholzer) [reply include/dd/Definitions.hpp:91]: Good suggestion! Changed it to `nextFP`.

File: include/dd/ComplexTable.hpp [hunk 3]
```diff
@@ -65,16 +66,25 @@ namespace dd {
...
+                return std::abs(e) <= TOLERANCE;
             }
 
-            [[nodiscard]] static inline bool approximatelyOne(const Entry* e) {
-                return e == &one || std::abs(val(e) - 1) < TOLERANCE;
+            [[nodiscard]] static constexpr bool approximatelyOne(const Entry* e) {
+                return e == &one || approximatelyOne(val(e));
+            }
+            [[nodiscard]] static constexpr bool approximatelyOne(fp e) {
+                return approximatelyEquals(e, 1.0);
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply include/dd/ComplexTable.hpp:87]: I'd rather keep it that way in order to benefit from having the same function (i.e. `approxiamatelyEquals`) used everywhere. Most compilers will inline the call anyway.

**Review by hillmich — APPROVED** [review]
LGTM

**Pull request merged**